## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [60]:
import pandas as pd
import numpy as np

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [61]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [62]:
# Use the describe function to gather some basic statistics
print(df.describe())

          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [63]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [64]:
# What three item categories had the most entries?
# Count the entries for each category
category_counts = df['category'].value_counts()

# Display the top 3 categories
print(category_counts.head(3))


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [65]:
# For the category with the most entries, which subcategory had the most entries?
top_category = category_counts.index[0]

# Filter the dataframe for the top category
top_category_df = df[df['category'] == top_category]

# Count the entries for each subcategory within the top category
subcategory_counts = top_category_df['subcategory'].value_counts()

# Display the top subcategory
print(f"For the category '{top_category}', the subcategory with the most entries is:")
print(subcategory_counts.head(1))


For the category 'consumables', the subcategory with the most entries is:
subcategory
bathroom supplies    6424
Name: count, dtype: int64


In [66]:
# Which five clients had the most entries in the data?
# Count the entries for each client
client_counts = df['client_id'].value_counts()

# Get the top 5 clients
top_5_clients = client_counts.head(5)
print("Top 5 clients by number of entries:")
print(top_5_clients)

Top 5 clients by number of entries:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [67]:
# Store the client ids of those top 5 clients in a list.
top_5_client_ids = top_5_clients.index.tolist()
print(top_5_client_ids)

[33615, 66037, 46820, 38378, 24741]


In [68]:
# How many total units (the qty column) did the client with the most entries order order?
# Calculate total units ordered by the client with the most entries
top_client_id = top_5_client_ids[0]
top_client_total_units = df[df['client_id'] == top_client_id]['qty'].sum()

print(f"\nThe client with the most entries (ID: {top_client_id}) ordered a total of {top_client_total_units} units.")


The client with the most entries (ID: 33615) ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [69]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'subtotal']].head())

   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [70]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping(weight):
    if weight > 50:
        return 7  # $7 per pound for orders over 50 pounds
    else:
        return 10  # $10 per pound for items 50 pounds or under

df['shipping_price'] = df.apply(lambda row: calculate_shipping(row['unit_weight'] * row['qty']) * row['unit_weight'] * row['qty'], axis=1)
print(df[['unit_price', 'unit_weight', 'qty', 'shipping_price']].head())

   unit_price  unit_weight  qty  shipping_price
0     1096.80         7.50  105         5512.50
1       24.95         1.49   21          312.90
2       13.52         1.68   39          458.64
3       36.42         1.23   29          356.70
4      195.10        46.43   20         6500.20


In [71]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = ((df['subtotal'] + df['shipping_price']) * 1.0925).round(2)
print(df[['subtotal', 'shipping_price', 'total_price']].head())

    subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40


In [72]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
print(df['line_cost'].head())

0    85597.05
1      629.79
2      765.18
3     1077.35
4     8663.60
Name: line_cost, dtype: float64


In [73]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
print(df['line_profit'].head())

0    46242.03
1      284.47
2      311.94
3      466.22
4     2700.80
Name: line_profit, dtype: float64


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [74]:
# Check your work using the totals above
# Function to get total price for a specific order
def get_order_total(order_id):
    order_total = df[df['order_id'] == order_id]['total_price'].sum()
    return round(order_total, 2)

# Check the three specific orders
order_ids = [2742071, 2173913, 6128929]
expected_totals = [152811.89, 162388.71, 923441.25]

for order_id, expected_total in zip(order_ids, expected_totals):
    calculated_total = get_order_total(order_id)
    print(f"Order ID {order_id}:")
    print(f"Calculated total: ${calculated_total:.2f}")
    print(f"Expected total: ${expected_total:.2f}")
    print(f"Difference: ${abs(calculated_total - expected_total):.2f}")
    print(f"Percentage difference: {abs(calculated_total - expected_total) / expected_total * 100:.4f}%")
    print()

# Display a few rows to verify the update
print(df[['subtotal', 'shipping_price', 'total_price']].head())

Order ID 2742071:
Calculated total: $152811.89
Expected total: $152811.89
Difference: $0.00
Percentage difference: 0.0000%

Order ID 2173913:
Calculated total: $162388.71
Expected total: $162388.71
Difference: $0.00
Percentage difference: 0.0000%

Order ID 6128929:
Calculated total: $923441.25
Expected total: $923441.25
Difference: $0.00
Percentage difference: 0.0000%

    subtotal  shipping_price  total_price
0  115164.00         5512.50    131839.08
1     523.95          312.90       914.26
2     527.28          458.64      1077.12
3    1056.18          356.70      1543.57
4    3902.00         6500.20     11364.40


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [81]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_clients = df['client_id'].value_counts().head(5).index.tolist()

client_spending = df[df['client_id'].isin(top_5_clients)].groupby('client_id')['total_price'].sum().sort_values(ascending=False)

print("Spending of top 5 clients by quantity:")
print(client_spending)
print()

Spending of top 5 clients by quantity:
client_id
24741    82268892.02
38378    12906550.87
66037    10259514.79
46820     9743794.36
33615     8377308.52
Name: total_price, dtype: float64



In [87]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_5_clients = df['client_id'].value_counts().head(5).index.tolist()
summary_df = df[df['client_id'].isin(top_5_clients)].groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'total_price': 'sum',
    'line_profit': 'sum'
}).rename(columns={
    'qty': 'Total Units',
    'shipping_price': 'Total Shipping',
    'total_price': 'Total Revenue',
    'line_profit': 'Total Profit'
})
summary_df.head()

Total Units  Total Shipping  Total Revenue  Total Profit
client_id                                                          
24741           239862      5126448.37    82268892.02   36579992.31
33615            64313      1828984.89     8377308.52    2201994.61
38378            73667      3429455.40    12906550.87    3271829.89
46820            75768      1601448.84     9743794.36    2736602.72
66037            43018      1395151.85    10259514.79    3255031.81

In [88]:
# Define the money columns
money_columns = ['Total Shipping', 'Total Revenue', 'Total Profit']

# Define a function that converts a dollar amount to millions
def currency_format_millions(value):
    return f"${value / 1e6:.2f}M"

# Apply the currency_format_millions function to only the money columns
for col in money_columns:
    summary_df[col] = summary_df[col].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format
summary_df = summary_df.rename(columns={
    'Total Shipping': 'Total Shipping (Millions)',
    'Total Revenue': 'Total Revenue (Millions)',
    'Total Profit': 'Total Profit (Millions)'
})
summary_df.head()

Total Units Total Shipping (Millions) Total Revenue (Millions)  \
client_id                                                                   
24741           239862                    $5.13M                  $82.27M   
33615            64313                    $1.83M                   $8.38M   
38378            73667                    $3.43M                  $12.91M   
46820            75768                    $1.60M                   $9.74M   
66037            43018                    $1.40M                  $10.26M   

          Total Profit (Millions)  
client_id                          
24741                     $36.58M  
33615                      $2.20M  
38378                      $3.27M  
46820                      $2.74M  
66037                      $3.26M

In [86]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
summary_df = summary_df.sort_values('Total Profit (Millions)', ascending=False)

print("Formatted summary for top 5 clients:")
print(summary_df)

Formatted summary for top 5 clients:
           Total Units Total Shipping (Millions) Total Revenue (Millions)  \
client_id                                                                   
24741           239862                    $5.13M                  $82.27M   
38378            73667                    $3.43M                  $12.91M   
66037            43018                    $1.40M                  $10.26M   
46820            75768                    $1.60M                   $9.74M   
33615            64313                    $1.83M                   $8.38M   

          Total Profit (Millions)  
client_id                          
24741                     $36.58M  
38378                      $3.27M  
66037                      $3.26M  
46820                      $2.74M  
33615                      $2.20M  


In [89]:
# Calculate some key metrics
total_profit = df['line_profit'].sum()
total_revenue = df['total_price'].sum()
top_5_profit = summary_df['Total Profit (Millions)'].apply(lambda x: float(x.replace('$', '').replace('M', ''))).sum() * 1e6
top_5_revenue = summary_df['Total Revenue (Millions)'].apply(lambda x: float(x.replace('$', '').replace('M', ''))).sum() * 1e6

# Write a brief summary
print("Summary of Findings:")
print(f"The top 5 clients by quantity account for {top_5_profit/total_profit:.2%} of total profit")
print(f"and {top_5_revenue/total_revenue:.2%} of total revenue.")
print(f"This concentration of business in a small number of clients suggests a potential")
print(f"risk if these key accounts were to be lost, but also presents opportunities for")
print(f"targeted relationship management and upselling to these high-value customers.")

Summary of Findings:
The top 5 clients by quantity account for 2.99% of total profit
and 2.19% of total revenue.
This concentration of business in a small number of clients suggests a potential
risk if these key accounts were to be lost, but also presents opportunities for
targeted relationship management and upselling to these high-value customers.
